### ISSUE TITLE
Replicas may be rescheduled even if `Failover` is disabled

### ISSUE THREAD

**Issue by Garrybest** [karmada-io/karmada#1640]
**What happened**:

In some circumstance even if `Failover` is disabled, replicas in some member cluster still have a change to be removed especially when scaling up. 

E.g.,

1. We have a deployment with a 5 desired replicas which are propagated to member cluster A.
2. The cluster is a little bit busy so the connection between karmada-agent and kube-apiserber of member A is lost for about 1 minute. So the cluster status is unhealthy.
3. It happens that the deployment has been scaled up to 10 replicas at this time.
4. Even if we disable Failover, the scaling up triggers the scheduling procedure, now the karmada-scheduler will remove replicas in cluster A which may be dangerous. However, it is possible that there is nothing wrong with the 5 replicas in member cluster A.

**What you expected to happen**:
We should not delete the replicas in unhealthy cluster.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Karmada version:
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
- Others:

**Comment by dddddai** [comment]
I think the root cause is that we don't **retain clusters** when scheduling, which means whenever a schedule is triggered, it would remove "unfit" clusters from rb.

This can be dangerous not only when clusters become not ready, but also when there's any change making clusters "unfit" (e.g. label changes...), because they will be evicted in the next schedule

Not sure if https://github.com/karmada-io/karmada/pull/1487 would fix this

**Comment by duanmengkk** [comment]
Excuse me, what do you means by **retain clusters** ?

**Comment by dddddai** [comment]
Let's say there's a RB scheduled to [A, B], then B becomes unhealthy, it won't perform schedule as `Failover` is disabled
However, in the next schedule(due to replicas change), B will be evicted from the RB while we should **retain B** although it becomes "unfit"

**Comment by dddddai** [comment]
I was thinking [using a parameter("evict")](https://github.com/karmada-io/karmada/pull/1049/files#diff-9c992787b9fc37779a22fd2f3700c60e272ab1b1ea3d89e91841a755f6727280R50) to indicate the scheduler whether to evict unfit clusters, do you have better ideas?

IIRC we should evict unfit clusters only when propagation policy changes, and then remove unhealthy clusters if `Failover` enabled

**Comment by duanmengkk** [comment]
> we should evict unfit clusters only when propagation policy changes

I am not quite sure.
if policy placement is
![图片](https://user-images.githubusercontent.com/17697155/165014983-d759f647-0721-4044-b6e7-77fea0c44bea.png)
when i changes the policy placement to
![图片](https://user-images.githubusercontent.com/17697155/165015134-8c7f6639-6f95-4dfe-8d00-0b74cf603dc0.png)
When member1 is unfit,should the replicas in member1 will be evicted?

**Comment by dddddai** [comment]
Yeah it's tricky, I think we can focus on `Failover` first: don't evict unhealthy clusters from RB if `Failover` disabled

**Comment by dddddai** [comment]
Yeah it's tricky, I think we can focus on `Failover` first:

![image](https://user-images.githubusercontent.com/41563853/165026751-ff227d42-8d10-4d0f-b53b-8d3c2fc73d7d.png)

**Comment by Garrybest** [comment]
> IIRC we should evict unfit clusters only when propagation policy changes, and then remove unhealthy clusters if Failover enabled

Agreed. These two scenarios are different.

**Comment by huone1** [comment]
let me list the rescheduling scenario， the following is what I think now：
1. propagation policy changes
2. cluster is unhealthy when Failover is enabled
3. cluster is unhealthy when Failover is disabled
4. cluster is deleted
5. a new cluster which is fit to PP is added

**Comment by duanmengkk** [comment]
> let me list the rescheduling scenario， the following is what I think now：
> 
>     1. propagation policy changes
> 
>     2. cluster is unhealthy when Failover is enabled
> 
>     3. cluster is unhealthy when Failover is disabled
> 
>     4. cluster is deleted
> 
>     5. a new cluster which is fit to PP is added

It seems lost a scenario this issue mentions：7. replicas of workload changes

**Comment by duanmengkk** [comment]
By the way, do you think **descheduler** should be responsible for these scenario? @Garrybest

**Comment by Garrybest** [comment]
> By the way, do you think **descheduler** should be responsible for these scenario? @Garrybest

Sounds good. Descheduler is meant to evict pods for workload. Now it only focus on `unschedulable` pods. The failover replicas could be counted on. 

However, descheduler works frequently (default interval: 2 minutes). Any scenario associated with PP change should be handled by scheduler, not by scheduler. We could have a further discussion tomorrow.

**Comment by huone1** [comment]
> > By the way, do you think **descheduler** should be responsible for these scenario? @Garrybest
> 
> Sounds good. Descheduler is meant to evict pods for workload. Now it only focus on `unschedulable` pods. The failover replicas could be counted on.
> 
> However, descheduler works frequently (default interval: 2 minutes). Any scenario associated with PP change should be handled by scheduler, not by scheduler. We could have a further discussion tomorrow.

I think we can extend the scheduling strategy in PP to define the rescheduling info; different workload can config the different policy

**Comment by snowplayfire** [comment]
> let me list the rescheduling scenario， the following is what I think now：
> 
> 1. propagation policy changes
> 2. cluster is unhealthy when Failover is enabled
> 3. cluster is unhealthy when Failover is disabled
> 4. cluster is deleted
> 5. a new cluster which is fit to PP is added
> 6. replicas of workload changes when the scheduled clusters are healthy
> 7. replicas of workload changes when some scheduled clusters are unhealthy

when cluster becomes healthy from unhealthy, is it a rescheduling case? @huone1

**Comment by RainbowMango** [comment]
> when cluster becomes healthy from unhealthy, is it a rescheduling case? @huone1

I don't think so. If all workloads are running, they shouldn't be re-scheduled to other clusters by default.

**Comment by Garrybest** [comment]
Hey guys, I have some preliminary ideas about this issue. 

First let's focus why this bug occurs.

1. When dividing replicas, we remove the previous target clusters which are not selected by scheduler in this round. See [here](https://github.com/karmada-io/karmada/blob/f55e898b75a8fff42a3b627822613cb7a566b203/pkg/scheduler/core/division_algorithm.go#L37-L42).
2. All replicas not in selected clusters will be considered as need-to-rescheduled ones. So we trigger scale-up-scheduling for a second sheduling.

Actually, when a cluster is unhealthy, it's dangerous to delete all replicas in this cluster for. Maybe a lot of workloads will be affected. 

Therefore, I prefer not to let `karmada-scheduler` remove target clusters in `spec.clusters`. Learn from what kube-scheduler does, it only cares about scheduling a pod to a node while descheduler focuses on evict pods in undesired nodes based on different policy. Here is my plan.

1. Add a new controller `cluster_lifecycle_controller`

This controller is inspired by `node_lifecycle_controller` in kubernetes. It is responsible for remove previous target clusters in `spec.clusters` when a cluster have some situations, e.g., the cluster is deleted or the cluster is unhealthy(only taking effect when `Failover` is enabled). This function is pretty like when a node is unhealthy, `node_lifecycle_controller` evict pods by a specified eviction rate.

2. Stop removing previous target clusters in `karmada-scheduler` when dividing replicas.

The `karmada-scheduler` does not remove previous target clusters and only cares about adding new target clusters or adjust replicas. All eviction should be handled by other components, like `cluster_lifecycle_controller` or `descheduler`.

3. How to handle 'propagation policy changes'?

Actually, must we need to reschedule all workloads without considering all previous PP and result? I'm not sure about whether this behavior is proper. But now the behavior seems like a little bit arbitrary. If we decide to keep this action, we could delete all previous result in `spec.clusters` before dividing replicas.

**Comment by huone1** [comment]
I agreed with the point 1, 2. For the 3rd, add a new field in  propagation policy to define whether reschedule all workloads without considering all previous PP and result?

**Comment by Garrybest** [comment]
> I agreed with the point 1, 2. For the 3rd, add a new field in propagation policy to define whether reschedule all workloads without considering all previous PP and result?

It's a good way to fix this problem. But I'm a little concernd about whether this field lead to more complication and confusion.

Another way is that we make this behavior as a global argument and give this job to descheduler. The karmada-descheduler decide how to evict replicas when a pp is not proper. It's just very similar as the descheduler decide how to evict pods when based on many policy.
